### PR TITLE
Fix PydanticObjectId fields being parsed into str

### LINKED ISSUE #940: [BUG] `PydanticObjectId` isn't properly configured when deserializing from JSON
**Describe the bug**
Consider the following code:
```python
class A(BaseModel):
   id: PydanticObjectId

raw = '{"id": "5eb7cf5a86d9755df3a6c593"}'
deserialized = A.model_validate_json(raw)
```

One would expect `type(deserialized.id) == PydanticObjectId`, but it's actually deserialized to `str`.

**To Reproduce**
See example above.

**Expected behavior**
`type(deserialized.id)` should be `PydanticObjectId`.

**Additional context**
Note that if I do `A.model_validate(json.loads(raw))` the code works. This is potentially because in the [`__get_pydantic_core_schema__` override in `PydanticObjectId`](https://github.com/BeanieODM/beanie/blob/main/beanie/odm/fields.py#L146), `python_schema` includes a transformation of the id from `str` to `ObjectId`, while the `json_schema` doesn't and directly uses the built-in `str_schema`.

### PR DESCRIPTION
**Pull request by 07pepa**
full implementation of https://github.com/BeanieODM/beanie/pull/1054 that passes test and include new ones

Fixes #940

### CODE REVIEW

**Review by Riverfount — DISMISSED** [review]
Great Job.

File: beanie/odm/fields.py [hunk 1]
```diff
@@ -25,7 +25,9 @@
 
 from bson import DBRef, ObjectId
 from bson.errors import InvalidId
-from pydantic import BaseModel
+from pydantic import (
+    BaseModel,
```
Comment (adeelsohailahmed) [inline beanie/odm/fields.py:29]: Unnecessary trailing comma

**Review by adeelsohailahmed — CHANGES_REQUESTED** [review]
Add a descriptive title to your PR, please. This title is used in the changelog. Adding a vague title doesn't tell us what was changed.

File: beanie/odm/fields.py (hunk 1 above)
Comment (07pepa) [reply beanie/odm/fields.py:29]: IMHO this should be configured and  fixed by ruff on commit

**Comment by 07pepa** [comment]
> Add a descriptive title to your PR, please. This title is used in the changelog. Adding a vague title doesn't tell us what was changed.


I did not know

File: beanie/odm/fields.py (hunk 1 above)
Comment (07pepa) [reply beanie/odm/fields.py]: removed and this should be caught by ruff

Comment (adeelsohailahmed) [reply beanie/odm/fields.py]: > IMHO this should be configured and fixed by ruff on commit

Sounds good, man. Happy to review the PR for it.

**Review by adeelsohailahmed — APPROVED** [review]
Thanks for the changes. LGTM.

File: beanie/odm/fields.py (hunk 1 above)
Comment (07pepa) [reply beanie/odm/fields.py]: will do when i get time for it

**Review by staticxterm — APPROVED** [review]
Just a bit worried if this would break other serialization/deserialization use cases for when Document.id is not an instance of a PydanticObjectId or string.

Otherwise looks good. Tested this a bit and it worked for me...

**Comment by mg3146** [comment]
> Just a bit worried if this would break other serialization/deserialization use cases for when Document.id is not an instance of a PydanticObjectId or string.
> 
> 
> 
> Otherwise looks good. Tested this a bit and it worked for me...

Just so I understand, would that be the case where people explicitly set the field "id" for some other use outside of mongo object id?

**Review by Riverfount — APPROVED** [review]
LGTM

**Review by ayusshsoarav — APPROVED** [review]

File: beanie/odm/fields.py [hunk 2]
```diff
@@ -159,7 +158,16 @@ def __get_pydantic_core_schema__(
...
+                json_schema=plain_validator(
+                    cls.validate,
+                    metadata={
+                        "pydantic_js_input_core_schema": core_schema.str_schema(
+                            pattern="^[0-9a-f]{24}$",
+                            min_length=24,
+                            max_length=24,
+                        )
+                    },
+                ),
```
Comment (ADR-007) [inline beanie/odm/fields.py:170]: Thank you for your contribution!
Could you please add a test for the generated JSON Schema as well? 
Unfortunately, it is causing issues with FastAPI OpenAPI documentation generation:
```
Cannot generate a JsonSchema for core_schema.PlainValidatorFunctionSchema ({'type': 'with-info', 'function': <bound method PydanticObjectId.validate of <class 'beanie.odm.fields.PydanticObjectId'>>})
```

<details>
  <summary>Stack trace</summary>
  
  ```
  .venv/lib/python3.12/site-packages/fastapi/applications.py:981: in openapi
    self.openapi_schema = get_openapi(
.venv/lib/python3.12/site-packages/fastapi/openapi/utils.py:493: in get_openapi
    field_mapping, definitions = get_definitions(
.venv/lib/python3.12/site-packages/fastapi/_compat.py:231: in get_definitions
    field_mapping, definitions = schema_generator.generate_definitions(
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:361: in generate_definitions
    self.generate_inner(schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:565: in generate_inner
    json_schema = current_handler(schema)
.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py:38: in __call__
    return self.handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:540: in new_handler_func
    json_schema = js_modify_function(schema_or_field, current_handler)
.venv/lib/python3.12/site-packages/pydantic/main.py:728: in __get_pydantic_json_schema__
    return handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py:38: in __call__
    return self.handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:493: in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:1489: in model_schema
    json_schema = self.generate_inner(schema['schema'])
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:565: in generate_inner
    json_schema = current_handler(schema)
.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py:38: in __call__
    return self.handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:493: in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:1061: in function_before_schema
    return self.generate_inner(schema['schema'])
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:565: in generate_inner
    json_schema = current_handler(schema)
.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py:38: in __call__
    return self.handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:493: in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:1603: in model_fields_schema
    json_schema = self._named_required_fields_schema(named_required_fields)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:1393: in _named_required_fields_schema
    field_json_schema = self.generate_inner(field).copy()
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:565: in generate_inner
    json_schema = current_handler(schema)
.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py:38: in __call__
    return self.handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:493: in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:1461: in model_field_schema
    return self.generate_inner(schema['schema'])
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:565: in generate_inner
    json_schema = current_handler(schema)
.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py:38: in __call__
    return self.handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:540: in new_handler_func
    json_schema = js_modify_function(schema_or_field, current_handler)
.venv/lib/python3.12/site-packages/beanie/odm/fields.py:178: in __get_pydantic_json_schema__
    json_schema = handler(schema)
.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py:38: in __call__
    return self.handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:493: in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:1343: in json_or_python_schema
    return self.generate_inner(schema['json_schema'])
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:565: in generate_inner
    json_schema = current_handler(schema)
.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py:38: in __call__
    return self.handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:493: in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:1088: in function_plain_schema
    return self.handle_invalid_for_json_schema(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pydantic.json_schema.GenerateJsonSchema object at 0x13cbeea20>
schema = {'function': {'function': <bound method PydanticObjectId.validate of <class 'beanie.odm.fields.PydanticObjectId'>>, 't...e_schema': {'max_length': 24, 'min_length': 24, 'pattern': '^[0-9a-f]{24}$', 'type': 'str'}}, 'type': 'function-plain'}
error_info = "core_schema.PlainValidatorFunctionSchema ({'type': 'with-info', 'function': <bound method PydanticObjectId.validate of <class 'beanie.odm.fields.PydanticObjectId'>>})"

    def handle_invalid_for_json_schema(self, schema: CoreSchemaOrField, error_info: str) -> JsonSchemaValue:
>       raise PydanticInvalidForJsonSchema(f'Cannot generate a JsonSchema for {error_info}')
E       pydantic.errors.PydanticInvalidForJsonSchema: Cannot generate a JsonSchema for core_schema.PlainValidatorFunctionSchema ({'type': 'with-info', 'function': <bound method PydanticObjectId.validate of <class 'beanie.odm.fields.PydanticObjectId'>>})
E       
E       For further information visit https://errors.pydantic.dev/2.10/u/invalid-for-json-schema

.venv/lib/python3.12/site-packages/pydantic/json_schema.py:2279: PydanticInvalidForJsonSchema
  ```
  
</details>

Comment (staticxterm) [reply beanie/odm/fields.py:170]: Hi @ADR-007, this is a closed pull request.
This issue is being addressed in #1080.

Comment (ADR-007) [reply beanie/odm/fields.py:170]: Thank you!

**Comment by RitinaADM** [comment]
Hello, friends. I'm not very familiar with GitHUB, have you found a solution to the problem yet?
FastAPI Users Beanie modules also crashed

**Pull request merged**